### PR TITLE
Adding a fixed version of dataflow dynamic rebalancing.

### PR DESCRIPTION
**Pull request by sduskis**
Dynamic rebalancing is Dataflow's ability to split long running, straggler Read jobs into 2 jobs.  That allows Dataflow to reduce overall execution time by moving parts of a job that are on busy machines to idle machines.

### CODE REVIEW

File: bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java [hunk 1]
```diff
@@ -235,12 +244,18 @@ public ResultArrayIterator(int arraySize) {
...
+    public Result[] next(ResultScanner<Row> resultScanner, ByteKeyRangeTracker rangeTracker)
         throws IOException {
       List<Result> results = new ArrayList<>();
       for (int i = 0; i < arraySize; i++) {
         Row row = resultScanner.next();
         if (row == null) {
+          // The scan completed.
+          break;
+        }
+        ByteKey key = ByteKey.copyFrom(row.getKey().toByteArray());
```
Comment (garye) [inline bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java]: Not sure what to do about it but this seems to make 2 array copies for every row which isn't ideal.

File: bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java [hunk 2]
```diff
@@ -745,17 +756,61 @@ public boolean advance() throws IOException {
...
+      return rangeTracker.getFractionConsumed();
+    }
+
+    /**
+     * Attempt to split the work by some percent of the ByteKeyRange based on a lexicographical
+     * split (and not statistics about the underlying table, which would be better, but that
+     * information does not exist).
+     */
+    @Override
+    public final synchronized BoundedSource<Results> splitAtFraction(double fraction) {
```
Comment (garye) [inline bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java]: can you add a unit test for this?

File: bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java [hunk 3]
```diff
@@ -745,17 +756,61 @@ public boolean advance() throws IOException {
...
+        estimatedSizeBytes = source.getEstimatedSizeBytes(null);
+      } catch (IOException e) {
+        READER_LOG.info("%s: Failed to get estimated size for key for fraction %s.", range, fraction);
+        return null;
+      }
+      long newEstimatedSize = (long) (fraction * estimatedSizeBytes);
+      byte[] splitKeyBytes = splitKey.getBytes();
+      SourceWithKeys<Results> residual = source.createSourceWithKeys(splitKeyBytes,
+        range.getEndKey().getBytes(), estimatedSizeBytes - newEstimatedSize);
+      this.source = source.createSourceWithKeys(range.getStartKey().getBytes(), splitKeyBytes,
```
Comment (garye) [inline bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java]: is this.source same as source? why refer to them differently?

**Comment by garye** [comment]
Would be helpful if the PR had some explanation of what "dataflow dynamic rebalancing" means.

File: bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java (hunk 1 above)
Comment (sduskis) [inline bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java]: There's a mismatch between our implementation of ByteString (which is shaded and renamed) and Dataflow's default package ByteString.  That's why the copying happens.  I can create a copy of the ByteKey, ByteKeyRange and ByteKeyRangeTracker classes from Dataflow that use our shaded ByteString.  That's what would be required to remove this copying.

Comment (garye) [inline bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java]: Let's not introduce anything like that until we can convince ourselves this is an actual problem.

**Comment by garye** [comment]
LGTM

**Comment by sduskis** [comment]
Closing this due to merge conflicts.  I created another PR for this.

**Pull request closed without merging**